### ISSUE TITLE
Add a Github specific reporter and enable it automatically by default based on environment variables

### ISSUE THREAD

**Issue by natebosch** [dart-lang/test#1415]
The default compact reporter is too noisy on github which does not support the line overwriting ascii escapes.

We should make a reporter which is less active, and consider adding support for the specific output format which allows the github UI to overlay failures on the code.

**Comment by kevmoo** [comment]
I did make a stab at this...nothing much to show of it, though https://github.com/dart-lang/test/tree/action_fun_null_safe

**Comment by devoncarew** [comment]
I did some experimentation with the github actions output options here:

https://github.com/devoncarew/output_tester/issues/2#issuecomment-1069826373

tldr is that there are some net improvements that can be made to the output, but the workflow commands are slightly less expressive overall than you'd want.

<img width="513" alt="Screen Shot 2022-03-17 at 9 30 11 AM" src="https://user-images.githubusercontent.com/1269969/158848776-1bcb22fc-a521-4a78-8c0d-5b1a87c5422e.png">

as compared to something like:

```
Run dart test

00:00 +0: loading test/output_tester_test.dart                                                                                                                                                         
00:01 +0: loading test/output_tester_test.dart                                                                                                                                                         
00:01 +0: test/output_tester_test.dart: calculate.1                                                                                                                                                    
00:01 +1: test/output_tester_test.dart: calculate.1                                                                                                                                                    
00:01 +1: test/output_tester_test.dart: calculate.2                                                                                                                                                    
00:01 +1 -1: test/output_tester_test.dart: calculate.2 [E]                                                                                                                                             
  Expected: <41>
    Actual: <42>
  
  package:test_api                   expect
  test/output_tester_test.dart 10:5  main.<fn>
  test/output_tester_test.dart [9](https://github.com/devoncarew/output_tester/pull/1/checks#step:8:9):[23](https://github.com/devoncarew/output_tester/pull/1/checks#step:8:23)  main.<fn>
  

00:01 +1 -1: test/output_tester_test.dart: group.a calculate.3                                                                                                                                         
00:01 +2 -1: test/output_tester_test.dart: group.a calculate.3                                                                                                                                         
00:01 +2 -1: test/output_tester_test.dart: group.a calculate.4                                                                                                                                         
00:02 +2 -1: test/output_tester_test.dart: group.a calculate.4                                                                                                                                         
00:02 +3 -1: test/output_tester_test.dart: group.a calculate.4                                                                                                                                         
00:02 +3 -1: test/output_tester_test.dart: calculate.2                                                                                                                                                 
00:02 +3 -2: test/output_tester_test.dart: calculate.2 [E]                                                                                                                                             
  Expected: <41>
    Actual: <42>
  
  package:test_api                    expect
  test/output_tester_test.dart 26:5   main.<fn>
  test/output_tester_test.dart [25](https://github.com/devoncarew/output_tester/pull/1/checks#step:8:25):23  main.<fn>
  

00:02 +3 -2: test/output_tester_test.dart: group.b calculate.3                                                                                                                                         
00:03 +3 -2: test/output_tester_test.dart: group.b calculate.3                                                                                                                                         
00:03 +4 -2: test/output_tester_test.dart: group.b calculate.3                                                                                                                                         
00:03 +4 -2: test/output_tester_test.dart: group.b calculate.4                                                                                                                                         
00:04 +4 -2: test/output_tester_test.dart: group.b calculate.4                                                                                                                                         
00:05 +4 -2: test/output_tester_test.dart: group.b calculate.4                                                                                                                                         
00:05 +5 -2: test/output_tester_test.dart: group.b calculate.4                                                                                                                                         
00:05 +5 -2: test/output_tester_test.dart: group.b calculate.5                                                                                                                                         
00:06 +5 -2: test/output_tester_test.dart: group.b calculate.5                                                                                                                                         
00:07 +5 -2: test/output_tester_test.dart: group.b calculate.5                                                                                                                                         
00:07 +6 -2: test/output_tester_test.dart: group.b calculate.5                                                                                                                                         
00:07 +6 -2: test/output_tester_test.dart: calculate.3                                                                                                                                                 
00:07 +6 -3: test/output_tester_test.dart: calculate.3 [E]                                                                                                                                             
  Expected: <[41](https://github.com/devoncarew/output_tester/pull/1/checks#step:8:41)>
    Actual: <[42](https://github.com/devoncarew/output_tester/pull/1/checks#step:8:42)>
  
  package:test_api                    expect
  test/output_tester_test.dart 50:5   main.<fn>
  test/output_tester_test.dart [49](https://github.com/devoncarew/output_tester/pull/1/checks#step:8:49):23  main.<fn>
  

00:07 +6 -3: Some tests failed.                                                                                                                                                                        

Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
Error: Process completed with exit code 1.
```

**Comment by jakemac53** [comment]
That definitely seems like an improvement 👍. Thanks for doing that exploration!

**Comment by kevmoo** [comment]
We also want to look at the auto-magic linking to the failure in the source
code. There's some magic for that, too.

This looks amazing, though!

**Comment by devoncarew** [comment]
Sure; I created a new PR in that example repo, with all the commits up through the one using the file markers;
https://github.com/devoncarew/output_tester/pull/3. You can see the some of their UI in the 'Run dart bin/output_tester.dart' step on the 'Checks' tab, and the UI contributed to the files in the 'Files changed' tab. Note that each file annotation is duplicated - I think that's because they are two PRs that had created these annotations, the file annotations feature is likely not used much, and looks buggy.

In the context of test output - when displaying a failure - I would see file markers being useful for
- showing a brief summary of the failure location, in-line with the rest of the test failure ('expectation foo failed, test/foo_test.dart, line 100')
- potentially have that summary being something you could use to jump to the failed test
- in the code for the failed test, having a brief inline summary, under the line for the expect() call ('expected 42; actual 41')

However:
- the visual summary for where you create a file marker has little useful info - no context for what it is
- 
<img width="452" alt="Screen Shot 2022-03-17 at 11 23 37 AM" src="https://user-images.githubusercontent.com/1269969/158870473-34db081d-1ca2-4d19-9918-9d68f74c9c8d.png">
- the in-line file marker is fairly large, w/ extra verbiage, and that hinders the user figuring out what it's there for
- 
<img width="554" alt="Screen Shot 2022-03-17 at 11 24 51 AM" src="https://user-images.githubusercontent.com/1269969/158870919-ac760ae5-171d-4c6f-a3a0-56ab61ebab22.png">
- the biggest drawback is that for the markers to be most useful, you have to be changing both the code under test (that causes the failure) and the test that fails, in the same PR. Otherwise there's no place on the PR to hang the file annotation off of. The Github UI appends a 'Unchanged files with check annotations' section to the PR for these annotations. These will be the norm however, and I don't think really help w/ understanding the failures.

I think the overall experience w/ file annotations is a bit odd, clunky, and just doesn't really line up well w/ what we want for test failure visualization.

**Comment by jakemac53** [comment]
> * the visual summary for where you create a file marker has little useful info - no context for what it is

So do these need an entire line to themselves or something, and can't be a part of the collapsed group?

> - the in-line file marker is fairly large, w/ extra verbiage, and that hinders the user figuring out what it's there for -

Ya it is large, although that would probably be acceptable, but also clicking it doesn't take you to the correct spot in the log... :(.

In general ya the feature doesn't really seem to be ready yet. Not sure if we should file some issues on them?

**Comment by kevmoo** [comment]
We could always have configuration and/or flags for the in-line file
markers. Some folks might like 'em

**Comment by devoncarew** [comment]
> So do these need an entire line to themselves or something, and can't be a part of the collapsed group?

I think they can be part of a collapsed group - I believe I may have experimented with that?

> In general ya the feature doesn't really seem to be ready yet. Not sure if we should file some issues on them?

For sure - there are a few things that could really help us out presentation-wise in the actions output. Some tweaks to file annotations, the ability to use colors in collapsed groups, perhaps supporting markdown processing, ...

**Comment by simolus3** [comment]
The new reporter, which is enabled by default in GitHub actions now, provides no information on tests failing if they fail due to an error after the main test has completed. Of course, this should never happen for well-written tests, but finding a bug in finalizers doesn't get easier with the reporter acting as if everything were ok :D

Both of these runs fail, but [this one](https://github.com/simolus3/sqlite3.dart/runs/8004749657?check_suite_focus=true) generates nothing but checkmarks before complaining about a failed test in the end. Re-running the [same tests](https://github.com/simolus3/sqlite3.dart/runs/8005394619?check_suite_focus=true) with the `expanded` reporter prints the proper error message. Could this information also be added to the default reporter?

**Comment by kevmoo** [comment]
@devoncarew – thoughts on ☝️

**Comment by devoncarew** [comment]
@simolus3 - the issue is that the github reporter has already written out its info, and thinks it is complete, but later gets new info (a failure)?

Can you re-file your above comment as a new issue? I agree this is something we should address. I agree we should output the information somehow, even if it doesn't line up well w/ the previously produced github actions formatted output.

**Comment by devoncarew** [comment]
I noticed that this reported is not exposed through `flutter test`; I filed https://github.com/flutter/flutter/issues/114860 to expose it.
